pkg/collector/corechecks/windows_event_log/bookmark.go

```diff
@@ -0,0 +1,76 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+//go:build windows
+
+package evtlog
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/DataDog/datadog-agent/pkg/util/winutil/eventlog/api"
+	"github.com/DataDog/datadog-agent/pkg/util/winutil/eventlog/bookmark"
+)
+
+type bookmarkSaver struct {
+	// inputs
+	bookmark          evtbookmark.Bookmark
+	bookmarkFrequency int
+	saveBookmark      func(bookmarkXML string) error
+
+	// track how often to save
+	eventsSinceLastBookmark int
+	lastBookmark            string
+	mu                      sync.Mutex
+}
+
+func (b *bookmarkSaver) updateBookmark(event *evtapi.EventRecord) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Updating and rendering the bookmark is fast, and it makes the "update bookmark at end of check"
+	// logic easier by avoiding having to track/save/close the event handle, so just do it every time.
+	err := b.bookmark.Update(event.EventRecordHandle)
+	if err != nil {
+		return fmt.Errorf("failed to update bookmark: %w", err)
+	}
+
+	bookmarkXML, err := b.bookmark.Render()
+	if err != nil {
+		return fmt.Errorf("failed to render bookmark XML: %w", err)
+	}
+	b.lastBookmark = bookmarkXML
```

Comment:
I tried to explain the reasoning in the comment just above
```
// Updating and rendering the bookmark is fast, and it makes the "update bookmark at end of check"
// logic easier by avoiding having to track/save/close the event handle, so just do it every time.
```

we want to save the bookmark [when the loop exits](https://github.com/DataDog/datadog-agent/blob/7bf529005ba4d724a242d64c1e130a72360dd722/pkg/collector/corechecks/windows_event_log/check.go#L96-L102), too.

and we also want to save it [when the check runs](https://github.com/DataDog/datadog-agent/blob/7bf529005ba4d724a242d64c1e130a72360dd722/pkg/collector/corechecks/windows_event_log/check.go#L76-L84)

If we don't render the XML, then we have to keep the event handle around (instead of closing it) so that we can render it when we do want to save it. And we have to be sure to keep that in sync with the subscription handle so we don't close the subscription and then try to use the (now defunct) event handle. event log handles are not supported by [DuplicateHandle](https://learn.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-duplicatehandle).

I tried doing that originally but the code wasn't neat, so I settled on this. The [log tailer](https://github.com/DataDog/datadog-agent/blob/28c8b70ace1c476cb5b1edcaaea6db061eb4f97e/pkg/logs/tailers/windowsevent/tailer.go#L323-L331) does the same thing for a similar reason (the logs agent handles the persistence automatically when the log is confirmed to have been sent).